Add support for regions in the C++ API

shado23 opened the pull request:
I happened to have a use case where I needed access to the raw pixel data through regions and discovered that the C++ API was missing support for them. 

I've quickly wired up a VRegion wrapper class based on the existing ones, and added a "region" method on VImage.

jcupitt commented:
Hey, this is great, thank you for doing this work. There are some very minor layout things, but we can fix them later. 

Could you also update `Doxyfile.in`? It's not generating docs for the new class yet. We'll need a one-line note in the main ChangeLog too (put your name on if you like).

I suppose we could have something to get `valid` and a reference to the underlying image. Perhaps that can wait for another PR.

kleisauke commented:
FWIW, this would collide with PR https://github.com/libvips/libvips/pull/1765.

jcupitt commented:
I didn't like the templates in 1765 :( perhaps that's the C programmer in me still resisting. I prefer the (imo, simpler) way you've done it here.

I missed that the 1765 author had updated again, I feel bad for not having persevered with that PR :(

shado23 commented:
> I didn't like the templates in 1765 :( perhaps that's the C programmer in me still resisting. I prefer the (imo, simpler) way you've done it here.

Hmm, yeah. I may be missing something, but I'm not sure what the templates would buy us here. There's only ever going to be one valid pixel format type for a given image, right? It might be worth it if there was a way to automagically cast the value to the correct format according to VipsBandFormat in typesafe way, but I'm not sure C++ can do that.

In any case, I've now added the missing methods that the previous PR had, added documentation, moved the trivial functions to the header, and made some minor tweaks to the spacing to better match the existing code style.

Let me know if there's anything else you would like changed.

shado23 commented:
I had accidentally made a mixed tab and spaces indentation mess, but that should be sorted now.

shado23 commented:
A conflict for Doxyfile.in was introduced by a recent commit. I've rebased the PR again, and it should be otherwise good to go unless there are any further changes to be made.

kleisauke commented on cplusplus/include/vips/VRegion8.h:
```diff
@@ -0,0 +1,146 @@
+	VipsPel *
+	addr( size_t i ) const
+	{
+		return &VIPS_REGION_ADDR_TOPLEFT( get_region() )[i];
```
I think this is more readable if one needs the address (there's no functional difference):
```suggestion
		return VIPS_REGION_ADDR_TOPLEFT( get_region() ) + i;
```

jcupitt commented on cplusplus/Doxyfile.in:
```diff
@@ -829,6 +829,7 @@ INPUT                  = \
 	@DOXY_INPUT_DIRECTORY@/VImage.cpp \
 	@DOXY_INPUT_DIRECTORY@/VInterpolate.cpp \
+	@DOXY_INPUT_DIRECTORY@/VRegion.cpp \
 	@DOXY_INPUT_DIRECTORY@/vips-operators.cpp \
```
I think we'll need `VRegion8.h` as well.

jcupitt commented on cplusplus/include/vips/VImage8.h:
```diff
@@ -1055,6 +1056,24 @@ class VImage : public VObject
 
+	/**
+	 * Acquire VRegion covering the whole image.
+	 */
```
This will make a region with no left/top/width/height set yet, rather than one that covers everything (that would need a huge amount of memory!).

jcupitt commented on cplusplus/include/vips/VRegion8.h:
```diff
@@ -0,0 +1,146 @@
+	operator()( int x, int y ) const
+	{
+		return *addr( x, y );
+	}
```
One other useful thing to expose would be `lskip` -- the region stride. Regions don't always have adjacent scanlines, so you need to add a specific `size_t` to move down one line.

Maybe (untested):

```C++
size_t
lskip()
{
    return VIPS_REGION_LSKIP( get_region() );
}
```

jcupitt commented on cplusplus/Makefile.am:
```diff
@@ -11,6 +11,7 @@ lib_LTLIBRARIES = libvips-cpp.la
 libvips_cpp_la_SOURCES = \
 	VImage.cpp \
 	VInterpolate.cpp \
+	VRegion.cpp \
```
`cplusplus/include/vips/Makefile.am` will need this too.

jcupitt commented on cplusplus/meson.build:
```diff
@@ -3,6 +3,7 @@ subdir('include/vips')
 libvips_cpp_lib = library('vips-cpp',
     'VImage.cpp',
     'VInterpolate.cpp',
+    'VRegion.cpp',
```
And a line in `include/vips/meson.build` for the header.

jcupitt reviewed (commented):
It looks great, just some small things.
